### PR TITLE
Add ClaimConfigInitDAO to allow extending the claim config creation

### PR DESCRIPTION
**Pull request by senthalan**
### Proposed changes in this pull request

- Add ClaimConfigInitDAO to allow extending the claim config creation

Part of https://github.com/wso2/product-is/issues/12535

### CODE REVIEW

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ClaimConfigInitDAO.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (janakamarasena) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ClaimConfigInitDAO.java:2]: Use new license header
```suggestion
 * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org).
```

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/internal/IdentityClaimManagementServiceComponent.java [hunk 2]
```diff
@@ -208,5 +214,37 @@ protected void unsetIdentityEventService(IdentityEventService identityEventServi
...
+            if (log.isDebugEnabled()) {
+                log.debug("Claim config init DAO implementation got registered: " +
+                        claimConfigInitDAO.getClass().getName());
+            }
+        }
+    }
+
+    protected void unsetClaimConfigInitDAO(ClaimConfigInitDAO claimConfigInitDAO) {
+
+        IdentityClaimManagementServiceDataHolder.getInstance().setClaimConfigInitDAO(new DefaultClaimConfigInitDAO());
```
Comment (janakamarasena) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/internal/IdentityClaimManagementServiceComponent.java:243]: When unsetting should we set if to default or completely remove it. If we need it shall we improve the debug log to mention the previous set dao was removed and now the default dao is set

Comment (senthalan) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/internal/IdentityClaimManagementServiceComponent.java:243]: we remove that DAO implementation and register the default DAO. I have improved the debug logs.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/internal/impl/DefaultClaimConfigInitDAO.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    @Override
+    public void initClaimConfig(ClaimConfig claimConfig, int tenantId) {
+
+        // Adding local claim dialect
+        try {
+            claimDialectDAO.addClaimDialect(new ClaimDialect(ClaimConstants.LOCAL_CLAIM_DIALECT_URI), tenantId);
+        } catch (ClaimMetadataException e) {
+            log.error("Error while adding claim dialect " + ClaimConstants.LOCAL_CLAIM_DIALECT_URI, e);
+        }
+
```
Comment (janakamarasena) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/internal/impl/DefaultClaimConfigInitDAO.java:62]: Additional new line

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/internal/impl/DefaultClaimConfigInitDAO.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        if (claimConfig.getClaimMap() != null) {
+
+            // Get the primary domain name
+            String primaryDomainName = IdentityUtil.getPrimaryDomainName();
+
+            // Adding external dialects and claims
+            Set<String> claimDialectList = new HashSet<>();
+
+            for (Map.Entry<ClaimKey, org.wso2.carbon.user.core.claim.ClaimMapping> entry : claimConfig.getClaimMap()
```
Comment (janakamarasena) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/internal/impl/DefaultClaimConfigInitDAO.java:72]: Shall we add an import for this? Didn't see any other `ClaimMapping` import

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/internal/impl/DefaultClaimConfigInitDAO.java (hunk 3 above)
Comment (senthalan) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/internal/impl/DefaultClaimConfigInitDAO.java:63]: removed

**Review by janakamarasena — APPROVED** [review]

**Pull request merged**